### PR TITLE
Support swapping the user context to the secondary auth user for named actions 

### PR DESCRIPTION
**Pull request by jakelandis**
This commit adds the ability to define specific actions that are required to be executed as the secondary authenticated user. When actions are defined as secondary auth actions, then the secondary authentication headers (and subsequent authentication) are required to call those actions and the authorization is based on the secondary user.  The SPI hook to define which actions are required is intended for internal only use only.

### CODE REVIEW

File: x-pack/plugin/security/qa/secondary-auth-actions/src/main/java/org/elasticsearch/secondary/auth/actions/SecondaryAuthActionsPlugin.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+package org.elasticsearch.secondary.auth.actions;
+
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.xpack.security.authc.support.SecondaryAuthActions;
+
+import java.util.Set;
+
+public class SecondaryAuthActionsPlugin extends Plugin implements SecondaryAuthActions {
+    public Set<String> get() {
+        return Set.of("cluster:admin/xpack/security/user/authenticate", "indices:admin/get");
```
Comment (jakelandis) [inline x-pack/plugin/security/qa/secondary-auth-actions/src/main/java/org/elasticsearch/secondary/auth/actions/SecondaryAuthActionsPlugin.java:17]: note - this is a test only plugin.

File: x-pack/plugin/security/qa/secondary-auth-actions/build.gradle [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  classname 'org.elasticsearch.secondary.auth.actions.SecondaryAuthActionsPlugin'
+  extendedPlugins = ['x-pack-security']
+}
+
+dependencies {
+  compileOnly project(':x-pack:plugin:core')
+  compileOnly project(':x-pack:plugin:security')
+  javaRestTestImplementation project(':test:framework')
+  // let the javaRestTest see the classpath of main
+  javaRestTestImplementation project.sourceSets.main.runtimeClasspath
```
Comment (mark-vieira) [inline x-pack/plugin/security/qa/secondary-auth-actions/build.gradle:16]: The preferred way to do this is `GradleUtils.extendSourceSet(project, 'main', 'javaRestTest')`

**Review by mark-vieira — CHANGES_REQUESTED** [review]

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java [hunk 3]
```diff
@@ -109,6 +114,16 @@ operations are blocked on license expiration. All data operations (read and writ
...
                 );
+            } else if (secondaryAuthActions.get().contains(action)) {
+                SecondaryAuthentication secondaryAuth = securityContext.getSecondaryAuthentication();
+                if (secondaryAuth == null) {
+                    throw new IllegalArgumentException("es-secondary-authorization header must be used to call action [" + action + "]");
+                } else {
+                    secondaryAuth.execute(ignore -> {
+                        applyInternal(task, chain, action, request, contextPreservingListener);
+                        return null;
+                    });
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java:125]: I think there's a problem here that this eats up the `_xpack_security_secondary_authc`.
A few actions rely on it to be present, see for eg `org.elasticsearch.xpack.transform.utils.SecondaryAuthorizationUtils`.

But, more importantly, the action filter is executed across the nodes.
I was able to make `SecondaryAuthActionIT` fail for the _get index_ action, which forwards the request to the master node, with:
```
diff --git a/x-pack/plugin/security/qa/secondary-auth-actions/src/javaRestTest/java/org/elasticsearch/secondary/auth/actions/SecondaryAuthActionsIT.java b/x-pack/plugin/security/qa/secondary-auth-actions/src/javaRestTest/java/org/elasticsearch/secondary/auth/actions/SecondaryAuthActionsIT.java
index 8f593d51bb1..e793912adfb 100644
--- a/x-pack/plugin/security/qa/secondary-auth-actions/src/javaRestTest/java/org/elasticsearch/secondary/auth/actions/SecondaryAuthActionsIT.java
+++ b/x-pack/plugin/security/qa/secondary-auth-actions/src/javaRestTest/java/org/elasticsearch/secondary/auth/actions/SecondaryAuthActionsIT.java
@@ -37,6 +37,7 @@ public class SecondaryAuthActionsIT extends ESRestTestCase {
     @ClassRule
     public static ElasticsearchCluster cluster = ElasticsearchCluster.local()
         .distribution(DistributionType.DEFAULT)
+        .nodes(2)
         .setting("xpack.watcher.enabled", "false")
         .setting("xpack.ml.enabled", "false")
         .setting("xpack.security.enabled", "true")
```

File: x-pack/plugin/security/qa/secondary-auth-actions/build.gradle [hunk 4]
```diff
@@ -12,8 +14,7 @@ dependencies {
   compileOnly project(':x-pack:plugin:core')
   compileOnly project(':x-pack:plugin:security')
   javaRestTestImplementation project(':test:framework')
-  // let the javaRestTest see the classpath of main
-  javaRestTestImplementation project.sourceSets.main.runtimeClasspath
+  GradleUtils.extendSourceSet(project, 'main', 'javaRestTest')
```
Comment (mark-vieira) [inline x-pack/plugin/security/qa/secondary-auth-actions/build.gradle:17]: nit: let's move this out of the `dependencies` configuration block

**Review by mark-vieira — APPROVED** [review]

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java (hunk 3 above)
Comment (jakelandis) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java]: thanks Albert !  

I fixed this in https://github.com/elastic/elasticsearch/pull/106613/commits/f3a4957e43be8dee6f51aa7103754e9f1d54406e

Comment (jakelandis) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java]: >A few actions rely on it to be present, see for eg org.elasticsearch.xpack.transform.utils.SecondaryAuthorizationUtils.

The actions names used here won't overlap with the pre-existing actions that use Secondary Auth already.

**Comment by jakelandis** [comment]
@elasticsearchmachine run elasticsearch-ci 
@elasticsearchmachine run elasticsearch-ci/part-1-fips 

(failure instance of https://github.com/elastic/elasticsearch/issues/106752)

**Review by albertzaharovits — APPROVED** [review]
LGTM

**Pull request merged**